include/boost/geometry/algorithms/detail/buffer/piece_border.hpp

```diff
@@ -0,0 +1,509 @@
+    lower = std::lower_bound(begin, end, value, compare);
+    // upper_bound points to first element of which value is LESS
+    upper = std::upper_bound(begin, end, value, compare);
+
+    if (lower != begin)
+    {
+        --lower;
+    }
+    if (upper != end)
+    {
+        ++upper;
+    }
+    return true;
+}
+
+}
+
+
+namespace detail { namespace buffer
+{
+
+//! Contains the border of the piece, consisting of 4 parts:
+//! 1: the part of the offsetted ring (referenced, not copied)
+//! 2: the part of the original (one or two points)
+//! 3: the left part (from original to offsetted)
+//! 4: the right part (from offsetted to original)
+//! Besides that, it contains some properties of the piece(border);
+//!   - convexity
+//!   - envelope
+//!   - monotonicity of the offsetted ring
+//!   - min/max radius of a point buffer
+//!   - if it is a "reversed" piece (linear features with partly negative buffers)
+template <typename Ring, typename Point>
+struct piece_border
+{
+    typedef typename geometry::coordinate_type<Point>::type coordinate_type;
+    typedef typename default_comparable_distance_result<Point>::type radius_type;
+    typedef typename geometry::strategy::buffer::turn_in_ring_winding<coordinate_type>::state_type state_type;
+
+    bool m_reversed;
+
+    // Points from the offsetted ring. They are not copied, this structure
+    // refers to those points
+    Ring const* m_ring;
+    std::size_t m_begin;
+    std::size_t m_end;
+
+    // Points from the original (one or two, depending on piece shape)
+    // Note, if there are 2 points, they are REVERSED w.r.t. the original
+    // Therefore here we can walk in its order.
+    boost::array<Point, 2> m_originals;
+    std::size_t m_original_size;
+
+    geometry::model::box<Point> m_envelope;
+    bool m_has_envelope;
+
+    // True if piece is determined as "convex"
+    bool m_is_convex;
+
+    // True if offsetted part is monotonically changing in x-direction
+    bool m_is_monotonic_increasing;
+    bool m_is_monotonic_decreasing;
+
+    radius_type m_min_comparable_radius;
+    radius_type m_max_comparable_radius;
+
+    piece_border()
+        : m_reversed(false)
+        , m_ring(NULL)
+        , m_begin(0)
+        , m_end(0)
+        , m_original_size(0)
+        , m_has_envelope(false)
+        , m_is_convex(false)
+        , m_is_monotonic_increasing(false)
+        , m_is_monotonic_decreasing(false)
+        , m_min_comparable_radius(0)
+        , m_max_comparable_radius(0)
+    {
+    }
+
+    //! Returns true if there is a ring, the ring has one or more points,
+    //! and the original has one or two points
+    bool valid() const
+    {
+        return m_ring != NULL && m_begin < m_end && m_original_size > 0;
+    }
+
+    // Only used for debugging (SVG)
+    Ring get_full_ring() const
+    {
+        Ring result;
+        if (! valid())
+        {
+            return result;
+        }
+        std::copy(m_ring->begin() + m_begin,
```

Comment:
This code should be enabled with an `#ifdef` because right now it requires `<iostream>` to be included which should not be the case.